packages/@aws-cdk/core/lib/private/uniqueid.ts

```diff
@@ -67,6 +68,10 @@ export function makeUniqueId(components: string[]) {
     .join('')
     .slice(0, MAX_HUMAN_LEN);
 
+  if (maxLength && human.length + hash.length > maxLength) {
+    return human.slice(0, maxLength - hash.length) + hash;
+  }
```

Comment:
This might be me being in way too many interviews, but would it not be nicer to find the longest duplicated substring in the string and remove the redundancy before the length checks?

`ApplicationStage-ApplicationStackApplicationPipelineStackApplicationStageApplicationStackApplicationClusterF3918940KubectlProvider5C2B31FB`
==>
`ApplicationStage-ApplicationStackApplicationPipelineStackApplicationStageClusterF3918940KubectlProvider5C2B31FB`

Reply:
OR, to swing the opposite way, would a map from word to abbreviation work?
Keeps all the semantic, but trims the length.

Reply:
We have to remove any redundancy after the length checks (otherwise, it's a breaking change). I'll update this PR to remove the longest duplicated substring if the length is larger than 128; if the resulting string is still longer, it will fall back to the logic here.